// rc-drawer/lib/Drawer.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread2"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var React = _interopRequireWildcard(require("react"));

var _PortalWrapper = _interopRequireDefault(require("rc-util/lib/PortalWrapper"));

var _DrawerPopup = _interopRequireDefault(require("./DrawerPopup"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var defaultGetContainer = function defaultGetContainer() {
  return document.body;
};

var Drawer = function Drawer(props) {
  var open = props.open,
      getContainer = props.getContainer,
      forceRender = props.forceRender,
      wrapperClassName = props.wrapperClassName,
      prefixCls = props.prefixCls,
      afterOpenChange = props.afterOpenChange,
      destroyOnClose = props.destroyOnClose;

  var _React$useState = React.useState(false),
      _React$useState2 = (0, _slicedToArray2.default)(_React$useState, 2),
      animatedVisible = _React$useState2[0],
      setAnimatedVisible = _React$useState2[1]; // ============================= Open =============================


  var internalAfterOpenChange = function internalAfterOpenChange(nextVisible) {
    setAnimatedVisible(nextVisible);
    afterOpenChange === null || afterOpenChange === void 0 ? void 0 : afterOpenChange(nextVisible);
  }; // ============================ Render ============================


  if (!forceRender && !animatedVisible && !open && destroyOnClose) {
    return null;
  }

  var sharedDrawerProps = (0, _objectSpread2.default)((0, _objectSpread2.default)({}, props), {}, {
    prefixCls: prefixCls,
    afterOpenChange: internalAfterOpenChange
  });

  if (getContainer === false) {
    return /*#__PURE__*/React.createElement(_DrawerPopup.default, (0, _extends2.default)({}, sharedDrawerProps, {
      inline: true
    }));
  }

  return /*#__PURE__*/React.createElement(_PortalWrapper.default, {
    visible: open,
    forceRender: forceRender,
    getContainer: getContainer,
    wrapperClassName: wrapperClassName
  }, function (_ref) {
    var scrollLocker = _ref.scrollLocker;
    return /*#__PURE__*/React.createElement(_DrawerPopup.default, (0, _extends2.default)({}, sharedDrawerProps, {
      scrollLocker: scrollLocker
    }));
  });
}; // Default Value.
// Since spread with default value will make this all over components.
// Let's maintain this in one place.


Drawer.defaultProps = {
  open: false,
  getContainer: defaultGetContainer,
  prefixCls: 'rc-drawer',
  placement: 'right',
  autoFocus: true,
  keyboard: true,
  width: 378,
  mask: true,
  maskClosable: true
};

if (process.env.NODE_ENV !== 'production') {
  Drawer.displayName = 'Drawer';
}

var _default = Drawer;
exports.default = _default;
